### PR TITLE
Curses: don't iterate through unbound COLORS #369

### PR DESCRIPTION
**Pull request by dankamongmen**
On DirectColor-capable terminals with the proper terminfo
database in use, COLORS is 2^24. Since the color map is
only 64k entries, this resulted in a segfault. I've
introduced NC::colorCount(), which bounds it by the
previously assumed maximum (and usable range) of 256.

### CODE REVIEW

File: src/curses/window.cpp [hunk 1]
```diff
@@ -224,9 +224,9 @@ int Color::pairNumber() const
 	else if (!isDefault())
 	{
 		if (!currentBackground())
-			result = (background() + 1) % COLORS;
+			result = (background() + 1) % colorCount();
 		result *= 256;
-		result += foreground() % COLORS;
+		result += foreground() % colorCount();
```
Comment (larson-carter) [inline src/curses/window.cpp]: Why did you move to colorCount() and not stick with COLORS? It seems like we would want to update COLORS instead of create a new method relying on colorCount()?? Since it is just an int being returned we could make it more efficient? Unless there is something that I am missing.

Comment (dankamongmen) [reply src/curses/window.cpp]: Since COLORS is owned by ncurses proper, I didn't want to touch it. Who knows what inside of ncurses is relying on this value? Who knows whether it might get changed (if for instance ncurses reread the terminfo definition)? The change was not made for efficiency (I very much doubt there's a measurable difference doing it one way or the other), but because we don't own `COLORS`. If you'd prefer it the other way (understandable also, since that way people don't need to know about `colorCount()`), I'm happy to change it up.

Comment (dankamongmen) [reply src/curses/window.cpp]: But yeah, from an efficiency perspective, this introduces a single function call to a non-hot path. If you're really worried about that, we can make it an inline in a header file, but I think such worry is misplaced. Let me know.

**Comment by dankamongmen** [comment]
BTW you might want to assign me #369

**Comment by dankamongmen** [comment]
If you'd prefer me to do it via assigning directly to `COLORS`, let me know, and I can change things up.

File: src/curses/window.cpp (hunk 1 above)
Comment (arybczak) [reply src/curses/window.cpp]: It's fine as it is, let's not microoptimize where it doesn't really matter.

**Comment by arybczak** [comment]
I don't think touching COLORS is a good idea, it belongs to ncurses.

**Comment by dankamongmen** [comment]
@arybczak sounds like we're in happy agreement, then. Can I get this merged? Thanks :).

**Comment by arybczak** [comment]
Now that I think about it, having a "global" local variable in window.cpp (i.e. in anonymous namespace) that is initialized in `initScreen` and then having `colorCount` simply return it would be cleaner. What do you say?

**Comment by dankamongmen** [comment]
I wanted to avoid anything with state because then we need to do it with `pthread_once_t()` to be threadsafe.

**Comment by arybczak** [comment]
`initScreen` is supposed to be called at the beginning of a program when no other threads are running, so it's fine.

**Comment by dankamongmen** [comment]
sure, change incoming

**Comment by dankamongmen** [comment]
PTAL

File: src/curses/window.cpp [hunk 2]
```diff
@@ -385,20 +385,34 @@ void disable()
 
 }
 
+// In a DirectColor setup, COLORS as returned by ncurses (via terminfo) can
+// run as high as 2^24. We only work with up to 256.
+int maxColor;
```
Comment (arybczak) [inline src/curses/window.cpp]: This should be inside anonymous namespace :)

**Comment by dankamongmen** [comment]
```
PRE[⎢] [63/67] 0x00 0x00 0x00 0x558cc8b1a3c0 depth: -3
NEEDMOVE: 0 INRIGHT: 0
POOPIN' IT
[⎢] [63/67] 0x00 0x00 0x00 0x558cc8b1a3c0
PRE[⎜] [63/68] 0x00 0x00 0x00 0x558cc8b1a3c0 depth: -3
NEEDMOVE: 0 INRIGHT: 1
PRE[⎷] [63/69] 0x00 0x00 0x00 0x558cc8b1a3c0 depth: -3
NEEDMOVE: 0 INRIGHT: 0
POOPIN' IT
[⎷] [63/69] 0x00 0x00 0x00 0x558cc8b1a3c0
PRE [ ] [63/70] 0x00 0x00 0x00 0x558cc8b1a3c0 depth: -3
NEEDMOVE: 0 INRIGHT: 1
PRE [c] [63/71] 0x00 0x00 0x00 0x558cc8b1a3c0 depth: -3
NEEDMOVE: 0 INRIGHT: 0
POOPIN' IT
```

why does it think we're in a right wide at 63/70?

**Comment by dankamongmen** [comment]
`[⎷] [63/69] WIDE: 1 0x00 0x00 0x00 0x55dab01b83c0` there's our problem!

**Comment by dankamongmen** [comment]
wait why are all the depths -3? the stdplane ought be -2, but it's coming in as -3.

**Comment by arybczak** [comment]
I think you're replying to the wrong thread :thinking:

File: src/curses/window.cpp (hunk 2 above)
Comment (dankamongmen) [reply src/curses/window.cpp]: Done.

**Comment by dankamongmen** [comment]
> I think you're replying to the wrong thread

indeed, thanks :) moved it into anonymous namespace

**Comment by arybczak** [comment]
Thanks!
